Update CI Page

KipHamiltons opened the pull request:
Extremely overdue. We got there. 

This is a page of all time.

https://deploy-preview-260--nubook.netlify.app/system/foundations/ci-system

KipHamiltons commented on src/book/02-system/01-foundations/03-continuous-integration.mdx:
```diff
@@ -39,155 +49,15 @@ This CI process works in combination with the [Git Feature Branch workflow](http
 
-If for example your PR build fails due to formatting errors, you can check the logs to see which files are incorrectly formatted. To fix the build, you would reformat the offending files and push a new commit, which will trigger a rebuild.
+Failed builds and pipeline steps can be diagnosed by viewing their logs.
+To view them, scroll to the build checks at the bottom of the PR page and click the **Details** link on the `buildkite/nubots/pr` check.
```
need to update this too

LachlanCourt commented on src/book/02-system/01-foundations/03-continuous-integration.mdx:
````diff
@@ -39,155 +49,31 @@ This CI process works in combination with the [Git Feature Branch workflow](http
-# Start the agent and configure it to run on system startup
-sudo systemctl enable buildkite-agent && sudo systemctl start buildkite-agent
-```
+Additionally, when pull requests are merged into main, docker images are built, tagged, and pushed to be the new standard images we use - for example, the `generic` image used for most of our development. This actions workflow is the primary way that those standard images are updated. They can be updated manually by pushing a local image, but this should only be done in exceptional circumstances.
````
We should document how to do this at some point but I think it's out of scope for this page

KipHamiltons replied:
Should we document that? I think it's as simple as `docker tag nubots/nubots:generic && docker push nubots/nubots:generic`, which is what you would expect, right? Is that something we need to cover in our docs?

In saying that, I feel like perhaps I know too much. Maybe I have lost touch with what nubook should have, since I am familiar with these tools now.

LachlanCourt replied:
Personally I didn't know how to do that although I would probably have been able to figure it out by looking though the docker documentation. I think it's a classic case of at what point does it change from documentation to tutorials. Probs fine to leave out tbh

LachlanCourt reviewed (commented):
Just some grammatical errors but otherwise looks great :heart_eyes:

KipHamiltons commented on src/book/02-system/01-foundations/03-continuous-integration.mdx:
```diff
@@ -8,26 +8,36 @@ authors:
-5. Buildkite receives the results and updates the build status in the Buildkite UI as well as on the PR page on GitHub.
-6. When the build passes, the PR is reviewed by another member of the team. If approved, the PR can be merged into `main`.
-7. Merging into `main` then triggers another CI build. This time, in addition to building the code, a separate job is created to rebuild the main image and push it to DockerHub. This job is handled by agents that have the [DockerHub login credentials](#add-dockerhub-credentials-to-buildkite), indicated by the `dockerhub=true` tag on the agent.
+2. Once development on the branch is complete, or reaches a point where it can be reviewed, the developer makes a pull request (PR) on GitHub. Creating the pull request triggers a GitHub action runners to check the code. GitHub automatically creates separate build runners for different parts of the PR (build, check code format, test, etc).
```
I might reword to "GitHub action agents" to keep it consistent with the other wording

KipHamiltons commented on src/book/02-system/01-foundations/03-continuous-integration.mdx:
```diff
@@ -39,155 +49,31 @@ This CI process works in combination with the [Git Feature Branch workflow](http
+When a pull request is created or its branch is modified, checks are run on the code to make sure that it builds and maintains standards of correctness. First, an action checks which parts of the codebase have been modified, which informs which checks should be run. For instance, changes to C++ code require the C++ code to be built, and its formatting checked.
 
-Steps from <https://buildkite.com/docs/agent/v3/ubuntu>.
+Next, a docker image tag is built and pushed to DockerHub for the PR. This is cached from the PR image, if it already exists, and the `main` branch image. The caching means that only what needs to be rebuilt in the image is rebuilt, to the point that if the image hasn't changed, the PR image tag simply acts as an alias for the image it cached from.
```
I agree this is definitely awkward wording. I think your suggestion works. There is the third case, that neither cache is valid for the changed dockerfile, in which case the thing is just built from the layer which diverges, but I don't think that's really an essential detail _we_ need to capture in our docs.

KipHamiltons commented:
Thanks for the review @LachlanCourt! Updated with your suggestions. 

I left a couple unresolved where I'm keen for your input.

When you re-review, note that I updated the glossary a bit as well.

LachlanCourt commented on src/book/03-guides/04-general/04-glossary.mdx:
```diff
@@ -80,6 +80,8 @@ authors:
 
 [**GitHub**](https://github.com/): A source code repository hosting service, which uses a git backend. NUbots uses GitHub to manage the source code.
 
+[**GitHub Actions**](https://docs.github.com/en/actions): A CI/CD cloud service and tools ecosystem with native integration with GitHub. NUbots uses GitHub actions to run CI builds and perform CD deployments for each code change. See our [section about our CI/CD setup](/system/foundations/ci-system).
```
```suggestion
[**GitHub Actions**](https://docs.github.com/en/actions): A CI/CD cloud service and tools ecosystem with native integration with GitHub. NUbots uses GitHub Actions to run CI builds and perform CD deployments for each code change. See our [section about our CI/CD setup](/system/foundations/ci-system).
```

LachlanCourt reviewed (commented):
Picky but for consistency 😬

Bidski commented on src/book/02-system/01-foundations/03-continuous-integration.mdx:
```diff
@@ -8,26 +8,36 @@ authors:
-6. When the build passes, the PR is reviewed by another member of the team. If approved, the PR can be merged into `main`.
-7. Merging into `main` then triggers another CI build. This time, in addition to building the code, a separate job is created to rebuild the main image and push it to DockerHub. This job is handled by agents that have the [DockerHub login credentials](#add-dockerhub-credentials-to-buildkite), indicated by the `dockerhub=true` tag on the agent.
+2. Once development on the branch is complete, or reaches a point where it can be reviewed, the developer makes a pull request (PR) on GitHub. Creating the pull request triggers GitHub action agents to check the code. GitHub automatically creates separate build agents for different parts of the PR (build, check code format, test, etc).
+3. Each agent completes a set of tasks specific in a pipeline file, found inside the [`.github/workflows`](https://github.com/NUbots/NUbots/tree/7cc0677eda6629c2b445a35473748c163685aaad/.github/workflows) directory of the codebase. These steps usually involve cloning the codebase, checking out the relevant commit, pulling down the docker image (rebuilding it if it has changed), then checking some ...
```
This link should reference the main branch
```suggestion
3. Each agent completes a set of tasks specified in a workflow file, found inside the [`.github/workflows`](https://github.com/NUbots/NUbots/tree/7cc0677eda6629c2b445a35473748c163685aaad/.github/workflows) directory of the codebase. These steps usually involve cloning the codebase, checking out the relevant commit, pulling down the docker image (rebuilding it if it has changed), then checking some aspect of the code.
```

Bidski commented on src/book/02-system/01-foundations/03-continuous-integration.mdx:
```diff
@@ -8,26 +8,36 @@ authors:
+2. Once development on the branch is complete, or reaches a point where it can be reviewed, the developer makes a pull request (PR) on GitHub. Creating the pull request triggers GitHub action agents to check the code. GitHub automatically creates separate build agents for different parts of the PR (build, check code format, test, etc).
+3. Each agent completes a set of tasks specific in a pipeline file, found inside the [`.github/workflows`](https://github.com/NUbots/NUbots/tree/7cc0677eda6629c2b445a35473748c163685aaad/.github/workflows) directory of the codebase. These steps usually involve cloning the codebase, checking out the relevant commit, pulling down the docker image (rebuilding it if it has changed), then checking some ...
+4. GitHub logs the results of each job and displays the results on the webpage for the pull request. The logs are viewable in the 'actions' tab of the PR.
+5. Once the build passes, the PR is reviewed by other members of the team. Once the PR has been approved, it can be merged into `main`.
```
```suggestion
5. Once the build passes, the PR is reviewed by other members of the team. Once the PR has been approved and all required checks pass, it can be merged into `main`.
```

Bidski commented on src/book/02-system/01-foundations/03-continuous-integration.mdx:
````diff
@@ -39,155 +49,31 @@ This CI process works in combination with the [Git Feature Branch workflow](http
-```
-
-### Install Docker Compose
+GitHub Actions are configured in [workflow files](https://docs.github.com/en/actions/using-workflows/workflow-syntax-for-github-actions), written in YAML. To add, remove, or change build jobs, we edit the relevant [workflow file in the NUbots repo](https://github.com/NUbots/NUbots/tree/7cc0677eda6629c2b445a35473748c163685aaad/.github/workflows). The [GitHub Actions documentation](https://docs.git ...
````
This link also need to reference the main branch

KipHamiltons replied:
permalinks are the good stuff

Bidski replied:
You want links to the commits rather than to main itself?

ysims replied:
+1 for permalinks but I'll grab a more recent commit at least. there's pros/cons to both but an out of date permalink is low risk low impact compare to a broken link imo.

KipHamiltons replied:
> You want links to the commits rather than to main itself?

If the path changes on main and you linked main, you have to start archaeology immediately to get any idea of the file contents. If you have a permalink, you can check main with a click, but you're guaranteed to get something useful even if it changed on main.

Bidski replied:
Sure, if you say so

KipHamiltons replied:
We take that as a W

Bidski commented on src/book/02-system/01-foundations/03-continuous-integration.mdx:
````diff
@@ -39,155 +49,31 @@ This CI process works in combination with the [Git Feature Branch workflow](http
-# Start the agent and configure it to run on system startup
-sudo systemctl enable buildkite-agent && sudo systemctl start buildkite-agent
-```
+Additionally, when pull requests are merged into main, docker images are built, tagged, and pushed to be the new standard images we use - for example, the `generic` image used for most of our development. This actions workflow is the primary way that those standard images are updated. They can be updated manually by pushing a local image, but this should only be done in exceptional circumstances.
````
I think either remove this last sentence or move it into a "details" section (you can then include basic instructions and a link to relevant external documentation if you use a details section). This is not an action that should be run (even if you manage to push broken/corrupted images to DockerHub you should still be able to fix them via a new PR) and the vast majority of people on the team don't even need to know that this is a possibility.
```suggestion
Additionally, when pull requests are merged into main, docker images are built, tagged, and pushed to be the new standard images we use - for example, the `generic` image used for most of our development. This actions workflow is the primary way that those standard images are updated.
```

JosephusPaye commented on src/book/02-system/01-foundations/03-continuous-integration.mdx:
````diff
@@ -39,155 +49,31 @@ This CI process works in combination with the [Git Feature Branch workflow](http
-```
-
-### Install Docker Compose
+GitHub Actions are configured in [workflow files](https://docs.github.com/en/actions/using-workflows/workflow-syntax-for-github-actions), written in YAML. To add, remove, or change build jobs, we edit the relevant [workflow file in the NUbots repo](https://github.com/NUbots/NUbots/tree/7cc0677eda6629c2b445a35473748c163685aaad/.github/workflows). The [GitHub Actions documentation](https://docs.git ...
````
Can't comment on the line directly but the image caption and file name can be updated to `main` instead of `master`.

JosephusPaye commented on src/book/02-system/01-foundations/03-continuous-integration.mdx:
```diff
@@ -8,26 +8,36 @@ authors:
+5. Once the build passes, the PR is reviewed by other members of the team. Once the PR has been approved and all required checks pass, it can be merged into `main`.
+6. Merging the PR into `main` then triggers another CI build. This time, in addition to building the code, a separate job is created to rebuild the main docker images and push them to DockerHub. This is the main way that our docker images are kept up to date.
 
 <TabbedImages>
```
Can't comment on the line directly but the image caption and file name can be updated to `main` instead of `master`.

KipHamiltons commented:
Lgtm, but the button is missing 🤪
